controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/StreamRequestHandler.java

```diff
@@ -99,8 +100,12 @@ public StreamRequestHandler(AutoScaleTask autoScaleTask,
     private <T extends ControllerEvent> CompletableFuture<Void> withCompletion(StreamTask<T> task,
                                                                                T event,
                                                                                Predicate<Throwable> writeBackPredicate) {
+        AtomicInteger i = new AtomicInteger(0);
         CompletableFuture<Void> result = new CompletableFuture<>();
-        task.execute(event)
+        Retry.withExpBackoff(100, 10, 6, 30000)
```

Comment:
We try for 71 seconds before throwing RetriesExhausted exception. Did not understand the reasoning behind it?

Reply:
So controller service receives a request, say seal stream. 
sealing a stream is a workflow that we need to perform. And once we have started it we want it to come to completion. we use event processor for running all our workflows. 

The workflow is broken into two phases.. phase 1 (first step) which is performed in the context of caller's api call. 
phase 2 which is performde on the event processor by picking the event. 

So when controllerService.streamMetadataTask.sealStream method is invoked, it first posts an event into requeststream followed by performing the processing of the first step of seal workflow (set state = sealing). 
If it fails to set the state to sealing, the call fails and we tell the caller that the request failed. 
In such situation we do not want to process the event that was posted. 
If it succeeds in setting the state to sealing, then the event processor should process the event.
Note: we cant perform first step and then post the event because we may fail before posting the event. 

However, posting of event triggers an asynchronous processing of the event. And this can race with caller's method invocation performing the first phase of the workflow. 
So if we pick the event but do not find first phase to have been performed, we retry a few times.. 1 minute or around 70 seconds is ample time for callers method to have performed the first step. If it is not performed within this time, it is assumed that the caller failed/crashed and the processing of the event is failed and ignored. 

There is a tradeoff in choosing the retry duration as all processing for a stream is serialized, so we introduce a delay equal to 71 seconds in case the caller crashed/failed which means we delay processing of any subsequent event for 71 seconds. 

There is an alternate way of achieving this same effect without this delay. But it will need some change to the infrastructure to move the event to the end of "in memory queue" that we maintain in SerializedRequestHandler everytime we encounter that first phase has not been performed. This way we can make sure we dont introduce delays in processing of subsequent events for the said stream and revisit this event and discard after a deterministic time if first phase has not been performed within that time.